clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -8601,12 +8601,23 @@ static void handleOpenCLAccessAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
   D->addAttr(::new (S.Context) OpenCLAccessAttr(S.Context, AL));
 }
 
-static void handleSYCLKernelAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+typedef void MangleContext::*Callback(const CXXRecordDecl *Lambda);
+static void handleSYCLKernelAttr(Sema &S, Decl *D, const ParsedAttr &AL, MangleContext &MC) {
   // The 'sycl_kernel' attribute applies only to function templates.
   const auto *FD = cast<FunctionDecl>(D);
   const FunctionTemplateDecl *FT = FD->getDescribedFunctionTemplate();
   assert(FT && "Function template is expected");
 
+  const CXXRecordDecl *Lambda;
+  if (KernelParamTy->isReferenceType())
+    Lambda = KernelParamTy->getPointeeCXXRecordDecl();
+  else
+    Lambda = KernelParamTy->getAsCXXRecordDecl();
+  QualType KernelNameType = FD->getTemplateSpecializationArgs()->get(0).getAsType();
+  SmallString<256> Result;
+  llvm::raw_svector_ostream Out(Result);
+
+  MC.mangleTypeName(KernelNameType, Lambda, MangleContext::markKernelNamingLambdas, Out);  
```

Comment:
I'm not certain this will work at this stage (I'm also not certain it *won't* work) -- the lambda you're trying to get a mangled name for may still be a dependent type, and so it might not be possible to mangle yet because it's not been instantiated. e.g.
```
auto lambda = [](auto i) [[clang::sycl_kernel]] { return i; }; // We process the attribute here

// But the auto parameter means the lambda is generic and so we don't know what
// its function signature is until we go to call the lambda.
```

Reply:
This is in an attempt to call the mangler early enough in the compilation in order to mark all the lambdas required to name it, but I see your point that this has potential to be too early. I will put this in a place that is a little bit later than this stage.